### PR TITLE
Convert handle to lower case before checking for uniqueness

### PR DESCRIPTION
**Pull request by F3Joule**

### CODE REVIEW

**Comment by F3Joule** [comment]
[Convert handle to lower case before checking for uniqueness and storing to Substrate storage](https://trello.com/c/m8HHjkqJ/363-convert-handle-to-lower-case-before-checking-for-uniqueness-and-storing-to-substrate-storage)

File: pallets/social/src/lib.rs [hunk 1]
```diff
@@ -24,7 +21,7 @@ pub struct Blog<T: Trait> {
 
   // Can be updated by the owner:
   pub writers: Vec<T::AccountId>,
-  pub slug: Vec<u8>,
+  pub handle: Vec<u8>,
```
Comment (siman) [inline pallets/social/src/lib.rs]: Hande should be optional.
```suggestion
  pub handle: Option<Vec<u8>>,
```

File: pallets/social/src/lib.rs [hunk 2]
```diff
@@ -38,7 +35,7 @@ pub struct Blog<T: Trait> {
 #[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug)]
 pub struct BlogUpdate<AccountId> {
   pub writers: Option<Vec<AccountId>>,
-  pub slug: Option<Vec<u8>>,
+  pub handle: Option<Vec<u8>>,
```
Comment (siman) [inline pallets/social/src/lib.rs]: Hande should be optional.
```suggestion
  pub handle: Option<Option<Vec<u8>>>,
```

File: pallets/social/src/lib.rs [hunk 3]
```diff
@@ -213,12 +210,14 @@ decl_error! {
...
-    /// Blog slug is not unique
-    SlugIsNotUnique,
+    /// Blog handle is too short
+    HandleIsTooShort,
+    /// Blog handle is too long
+    HandleIsTooLong,
+    /// Blog handle is not unique
+    HandleIsNotUnique,
+    /// Blog handle contains invalid characters
+    HandleContainsRestrictedChar,
```
Comment (siman) [inline pallets/social/src/lib.rs]:
```suggestion
    HandleContainsInvalidChars,
```

File: pallets/social/src/lib.rs [hunk 4]
```diff
@@ -430,12 +429,10 @@ decl_module! {
     // this is needed only if you are using events in your pallet
     fn deposit_event() = default;
 
-    pub fn create_blog(origin, slug: Vec<u8>, ipfs_hash: Vec<u8>) {
+    pub fn create_blog(origin, handle: Vec<u8>, ipfs_hash: Vec<u8>) {
```
Comment (siman) [inline pallets/social/src/lib.rs]:
```suggestion
    pub fn create_blog(origin, handle: Option<Vec<u8>>, ipfs_hash: Vec<u8>) {
```

File: pallets/social/src/lib.rs [hunk 5]
```diff
@@ -430,12 +429,10 @@ decl_module! {
...
       let owner = ensure_signed(origin)?;
 
-      ensure!(slug.len() >= Self::slug_min_len() as usize, Error::<T>::SlugIsTooShort);
-      ensure!(slug.len() <= Self::slug_max_len() as usize, Error::<T>::SlugIsTooLong);
-      ensure!(!BlogIdBySlug::exists(slug.clone()), Error::<T>::SlugIsNotUnique);
+      let handle = Self::if_blog_handle_valid_then_to_lower(handle.clone())?;
```
Comment (siman) [inline pallets/social/src/lib.rs]:
```suggestion
      let handle = Self::lowercase_and_validate_a_handle(handle.clone())?;
```

File: pallets/social/src/lib.rs [hunk 6]
```diff
@@ -456,7 +453,7 @@ decl_module! {
       Self::add_blog_follower_and_insert_blog(owner.clone(), new_blog, true)?;
 
       <BlogIdsByOwner<T>>::mutate(owner.clone(), |ids| ids.push(blog_id));
-      BlogIdBySlug::insert(slug, blog_id);
+      BlogIdByHandle::insert(handle, blog_id);
```
Comment (siman) [inline pallets/social/src/lib.rs]: Maybe wrap `blog_id` into `Some(blog_id)`?
```suggestion
      BlogIdByHandle::insert(handle, Some(blog_id));
```

File: pallets/social/src/lib.rs [hunk 7]
```diff
@@ -500,17 +497,14 @@ decl_module! {
...
-          ensure!(slug_len <= Self::slug_max_len() as usize, Error::<T>::SlugIsTooLong);
-          ensure!(!BlogIdBySlug::exists(slug.clone()), Error::<T>::SlugIsNotUnique);
-
-          BlogIdBySlug::remove(blog.slug.clone());
-          BlogIdBySlug::insert(slug.clone(), blog_id);
-          new_history_record.old_data.slug = Some(blog.slug);
-          blog.slug = slug;
+      if let Some(mut handle) = update.handle {
+        if handle != blog.handle {
+          handle = Self::if_blog_handle_valid_then_to_lower(handle.clone())?;
```
Comment (siman) [inline pallets/social/src/lib.rs]:
```suggestion
          handle = Self::lowercase_and_validate_a_handle(handle.clone())?;
```

File: pallets/social/src/tests.rs [hunk 8]
```diff
@@ -68,18 +68,18 @@ pub type AccountId = u64;
...
+fn blog_handle() -> Vec<u8> {
+  b"blog_handle".to_vec()
 }
 
 fn blog_ipfs_hash() -> Vec<u8> {
   b"QmRAQB6YaCyidP37UdDnjFY5vQuiBrcqdyoW1CuDgwxkD4".to_vec()
 }
 
-fn blog_update(writers: Option<Vec<AccountId>>, slug: Option<Vec<u8>>, ipfs_hash: Option<Vec<u8>>) -> BlogUpdate<u64> {
+fn blog_update(writers: Option<Vec<AccountId>>, handle: Option<Vec<u8>>, ipfs_hash: Option<Vec<u8>>) -> BlogUpdate<u64> {
```
Comment (siman) [inline pallets/social/src/tests.rs]:
```suggestion
fn blog_update(writers: Option<Vec<AccountId>>, handle: Option<Option<Vec<u8>>>, ipfs_hash: Option<Vec<u8>>) -> BlogUpdate<u64> {
```

File: pallets/social/src/tests.rs [hunk 9]
```diff
@@ -169,10 +169,10 @@ fn _create_default_blog() -> DispatchResult {
   _create_blog(None, None, None)
 }
 
-fn _create_blog(origin: Option<Origin>, slug: Option<Vec<u8>>, ipfs_hash: Option<Vec<u8>>) -> DispatchResult {
+fn _create_blog(origin: Option<Origin>, handle: Option<Vec<u8>>, ipfs_hash: Option<Vec<u8>>) -> DispatchResult {
```
Comment (siman) [inline pallets/social/src/tests.rs]:
```suggestion
fn _create_blog(origin: Option<Origin>, handle: Option<Option<Vec<u8>>>, ipfs_hash: Option<Vec<u8>>) -> DispatchResult {
```

File: pallets/social/src/functions.rs [hunk 10]
```diff
@@ -275,4 +276,24 @@ impl<T: Trait> Module<T> {
...
     }
+
+    fn is_valid_handle_char(c: u8) -> bool {
+        match c {
+            b'0'..=b'9' | b'a'..=b'z' | b'_' => true,
+            _ => false,
+        }
+    }
+
+    pub fn if_blog_handle_valid_then_to_lower(mut handle: Vec<u8>) -> Result<Vec<u8>, DispatchError> {
```
Comment (siman) [inline pallets/social/src/functions.rs]:
```suggestion
    pub fn lowercase_and_validate_a_handle(mut handle: Vec<u8>) -> Result<Vec<u8>, DispatchError> {
```

**Review by siman — CHANGES_REQUESTED** [review]

File: pallets/social/src/lib.rs (hunk 6 above)
Comment (F3Joule) [reply pallets/social/src/lib.rs]: The inner Option type is required in `::insert` by StorageMap

**Review by siman — APPROVED** [review]

**Pull request merged**